File: analysis/text_analysis.py
from collections import Counter
import os

def analyze_file(filename):
    path = os.path.join("datasets", filename)

    try:
        with open(path, "r", encoding="utf-8") as file:
            text = file.read()

        characters = len(text)
        words = len(text.split())
        lines = len(text.splitlines())

        unique_characters = len(set(text))

        letters = [c.lower() for c in text if c.isalpha()]
        frequency = Counter(letters)

        print("\n----- File Analysis -----")
        print(f"Characters       : {characters}")
        print(f"Words            : {words}")
        print(f"Lines            : {lines}")
        print(f"Unique Characters: {unique_characters}")

        print("\nLetter Frequency")
        for letter, count in sorted(frequency.items()):
            print(f"{letter} : {count}")

    except FileNotFoundError:
        print("File not found.")

File: analysis/text_analysis.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <map>
#include <set>

using namespace std;

void analyzeFile(string filename)
{
    ifstream file("datasets/" + filename);

    if (!file){
        cout<<"File not found."<<endl;
        return;
    }

    string line;
    int characters=0, words=0, lines=0;
    map<char, int> frequency;
    set<char> uniquechars;

    while (getline(file, line)) {

        lines++;
        characters+=line.length() ;

        string word = "";

        for (char c:line){

            uniquechars.insert(c);

            if (isalpha(c)) frequency[tolower(c)]++;

            if (isspace(c)) {

                if (word!="") {
                    words++;
                    word="";
                }
            }

            else{
                word+=c;
            }
        }

        if (word!="")
            words++;
    }

    cout<<"\n----- File Analysis -----"<<endl;
    cout<<"Characters       : " << characters << endl;
    cout<<"Words            : " << words << endl;
    cout<<"Lines            : " << lines << endl;
    cout<<"Unique Characters: " << uniquechars.size() << endl;

    cout << "\nLetter Frequency" << endl;
    for (auto x:frequency) {
        cout << x.first << " : " << x.second << endl;
    }

    file.close();
}

int main()
{
    string filename;
    cout << "Enter filename: ";
    cin >> filename;

    analyzeFile(filename);

    return 0;
}